### ISSUE TITLE
Using Threads in Interactive Systems: A Case Study

### ISSUE THREAD

**Issue by eric-haibin-lin** [eric-haibin-lin/computer-system-readings#1]
This pager examines the use of threads in real computer systems. Thread, a new abstraction 
back then, has different levels of priorities for scheduling, while conditional variable and monitor are both provided for thread activity synchronization.

Threads are used for different purposes, the most important ones are the following: deferred work, where work is done by the thread when the system is less busy; pump, where consumer and producer threads work together to achieve one task; sleepers and one shots, where the thread is doing few work and wakes up only once in a while; deadlock avoiders, task rejuvenation. etc. 

The thread abstraction, is easy to use if it’s for deadlock avoiders and pumps. However, it has potential issues when used to implement producer consumer threads for batched request processing. For example, without correct priority setup, the consumer thread may end up not merging any request. One the other hand, thread fork may fail due to resource starvation, and is not robust enough when the environment changes.
